libr/bin/dwarf.c

```diff
@@ -1319,321 +1519,523 @@ static void r_bin_dwarf_dump_debug_info(FILE *f, const RBinDwarfDebugInfo *inf)
+	case DW_FORM_ref_sig8:
+		value->reference = READ64 (buf);
+		break;
+	// offset into .debug_line_str section, can't parse the section now, so we just skip
+	case DW_FORM_strx:
+		buf = r_uleb128 (buf, buf_end - buf, &value->string.offset);
+		break;
+	case DW_FORM_strx1:
+		value->string.offset = READ8 (buf);
+		break;
+	case DW_FORM_strx2:
+		value->string.offset = READ16 (buf);
+		break;
+	case DW_FORM_strx3: // TODO Add 3 byte int read
+		buf += 3;
+		break;
+	case DW_FORM_strx4:
+		value->string.offset = READ32 (buf);
+		break;
+	case DW_FORM_implicit_const:
+		value->sdata = def->special;
+		break;
+	/*  addrx* forms : The index is relative to the value of the
+		DW_AT_addr_base attribute of the associated compilation unit. 
+	    index into an array of addresses in the .debug_addr section.*/
+	case DW_FORM_addrx:
+		buf = r_uleb128 (buf, buf_end - buf, &value->address);
+		break;
+	case DW_FORM_addrx1:
+		value->address = READ8 (buf);
+		break;
+	case DW_FORM_addrx2:
+		value->address = READ16 (buf);
+		break;
+	case DW_FORM_addrx3:
+	// I need to add 3byte endianess free read here TODO
+		buf += 3;
+		break;
+	case DW_FORM_addrx4:
+		value->address = READ32 (buf);
+		break;
+	case DW_FORM_line_ptr: // offset in a section .debug_line_str
+	case DW_FORM_strp_sup: // offset in a section .debug_line_str
+		value->string.offset = dwarf_read_piece(hdr->is_64bit, &buf, buf_end);
+		break;
+	// offset in the supplementary object file
+	case DW_FORM_ref_sup4:
+		value->reference = READ32 (buf);
+		break;
+	case DW_FORM_ref_sup8:
+		value->reference = READ64 (buf);
+		break;
+	// An index into the .debug_loclists
+	case DW_FORM_loclistx:
+	 // An index into the .debug_rnglists
+	case DW_FORM_rnglistx:
+		buf = r_uleb128 (buf, buf_end - buf, &value->address);
 		break;
 	default:
-		eprintf ("Unknown DW_FORM 0x%02"PFMT64x"\n", spec->attr_form);
-		value->encoding.data = 0;
+		eprintf ("Unknown DW_FORM 0x%02" PFMT64x "\n", def->attr_form);
+		value->data = 0;
 		return NULL;
 	}
 	return buf;
 }
 
-static const ut8 *r_bin_dwarf_parse_comp_unit(Sdb *s, const ut8 *obuf,
-		RBinDwarfCompUnit *cu, const RBinDwarfDebugAbbrev *da,
-		size_t offset, const ut8 *debug_str, size_t debug_str_len) {
-	const ut8 *buf = obuf, *buf_end = obuf + (cu->hdr.length - 7);
-	ut64 abbr_code;
-	size_t i;
+/**
+ * @brief
+ * 
+ * @param buf Start of the DIE data
+ * @param buf_end
+ * @param abbrev Abbreviation of the DIE
+ * @param hdr Unit header
+ * @param die DIE to store the parsed info into
+ * @param debug_str Ptr to string section start
+ * @param debug_str_len Length of the string section
+ * @param sdb 
+ * @return const ut8* Updated buffer
+ */
+static const ut8 *parse_die(const ut8 *buf, const ut8 *buf_end, RBinDwarfAbbrevDecl *abbrev, 
+		RBinDwarfCompUnitHdr *hdr, RBinDwarfDie *die, const ut8 *debug_str, size_t debug_str_len, Sdb *sdb) {
+
+	for (size_t i = 0; i < abbrev->length - 1; i++) {
+		if (die->length == die->capacity) {
+			expand_die (die);
```

Comment:
if the initialization of the `die` is done after getting the right abbreviation entry this should not be necessary, because the number of attributes is already known.

Reply:
yes, indeed